cluster-provision/tools/check-image-pull-policies/check-image-pull-policies.go

```diff
@@ -0,0 +1,177 @@
+	dryRun         bool
+}
+
+func flagOptions() options {
+	o := options{}
+	flag.StringVar(&o.manifestSource, "manifest-source", "", "The directory with manifest files or the manifest file to check. If a directory is given, all files are tried to parse as manifest")
+	flag.BoolVar(&o.dryRun, "dry-run", true, "Whether to exit with a non-zero exit code if the check fails")
+	flag.BoolVar(&o.verbose, "verbose", false, "Whether to output all parsed information or only the information on where the checks failed")
+	flag.Parse()
+	return o
+}
+
+var deserializer = serializer.NewCodecFactory(runtime.NewScheme()).UniversalDeserializer()
+
+func main() {
+	options := flagOptions()
+
+	if options.manifestSource == "" {
+		log.Fatal("No manifest-source given!")
+	}
+
+	fileInfo, err := os.Stat(options.manifestSource)
+	if os.IsNotExist(err) {
+		log.Fatalf("Failed to open %s: %v", options.manifestSource, err)
+	}
+
+	filesWithPullPolicies := map[string]map[string]corev1.PullPolicy{}
+	if fileInfo.IsDir() {
+		err = filepath.Walk(options.manifestSource, func(path string, info os.FileInfo, err error) error {
+			if info.IsDir() {
+				return nil
+			}
+			checkFileForPullPolicies(path, filesWithPullPolicies)
+			return nil
+		})
+		if err != nil {
+			log.Fatalf("Error on walking path %s: %v", options.manifestSource, err)
+		}
+	} else {
+		checkFileForPullPolicies(options.manifestSource, filesWithPullPolicies)
+	}
+
+	fmt.Printf("%d files with pull policies detected\n", len(filesWithPullPolicies))
+	hasOffendingPolicies := false
+	for filePath, pullPolicies := range filesWithPullPolicies {
+		for image, pullPolicy := range pullPolicies {
+			imageParts := strings.Split(image, ":")
+			var imageTag string
+			if len(imageParts) > 0 {
+				imageTag = imageParts[1]
+			}
+			offending := pullPolicy == corev1.PullAlways ||
+				(pullPolicy == "" && (imageTag == "" || imageTag == "latest"))
+			if offending {
+				hasOffendingPolicies = true
+			}
+			if !offending && !options.verbose {
+				continue
+			}
+			fmt.Printf("File: %s\n", filePath)
+			fmt.Printf("\tImage: %s\n", image)
+			// https://kubernetes.io/docs/concepts/containers/images/#updating-images
+			if offending {
+				fmt.Printf("\t\t-> PullPolicy: %s\n", pullPolicy)
+			} else {
+				fmt.Printf("\t\t   PullPolicy: %s\n", pullPolicy)
+			}
+		}
+	}
+	if hasOffendingPolicies {
+		if options.dryRun {
+			fmt.Println("WARNING: detected pull policies that will always pull images!")
+		} else {
+			log.Fatal("ERROR: detected pull policies that will always pull images!")
+		}
+	}
+}
+
+func checkFileForPullPolicies(manifestFile string, filesWithPullPolicies map[string]map[string]corev1.PullPolicy) {
+	file, err := os.Open(manifestFile)
+	if err != nil {
+		log.Fatalf("Error on opening file %s: %v", manifestFile, err)
+	}
+	//noinspection GoUnhandledErrorResult
+	defer file.Close()
+
+	pullPolicies := map[string]corev1.PullPolicy{}
+	scanner := bufio.NewScanner(file)
+	var bufferString *bytes.Buffer
```

Comment:
Maybe use a proper YAML parser like https://github.com/go-yaml/yaml?

Reply:
I don't understand? I'm using the k8s deserializer [here](https://github.com/kubevirt/kubevirtci/pull/407/files/9643c60e6ae92c1ee156a58cabefbb0af4a501a2#diff-1925687890c7a0e40494aa643f0b97c7R33)? What is bad about that?

Reply:
@qinqon @danielBelenky you might be aware that we need to scan files with multiple yaml documents of which we don't know what to expect there, right? Now, as we are using a proper k8s yaml decoder later on, I thin we are already good by just fetching just the documents one after another, no?

Reply:
I am ok with manual